File: FunctionalConcurrencyNET/Chapter.14/StockTicker.Server.Cs/Bus/CommandHandler.cs
using StockTicker.Server.Cs.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using static StockTicker.Core.Models;
using StockTicker.Events;
using static StockTicker.Events.Events;

namespace StockTicker.Server.Cs.Bus
{
    public static class CommandHandler
    {
        private static TradingCoordinator tradingCoordinator = TradingCoordinator.Instance();
        private static EventStorage.EventStorage eventStorage = new EventStorage.EventStorage();

        public static void Handle (CommandWrapper commandWrapper)
        {
            var connectionId = commandWrapper.ConnectionId;

            var command = TradingSupervisorAgent.CoordinatorMessage.NewPublishCommand(connectionId, commandWrapper);
            tradingCoordinator.PublishCommand(command);   // #E

            Event eventData;
            if (commandWrapper.Command is TradingCommand.BuyStockCommand cmdBuy)
            {
                eventData = Event.NewStocksBuyedEvent(commandWrapper.Id, cmdBuy.tradingRecord);
            }
            else if (commandWrapper.Command is TradingCommand.SellStockCommand cmdSell)
            {
                eventData = Event.NewStocksSoldEvent(commandWrapper.Id, cmdSell.tradingRecord);
            }
            else
            {
                throw new Exception("Unsupported command type");
            }

            var eventDescriptor = Event.Create(commandWrapper.Id, eventData);
            eventStorage.SaveEvent(Guid.Parse(connectionId), eventDescriptor);
        }
    }
}

File: FunctionalConcurrencyNET/Chapter.14/StockTicker.Server/Bus/CommandHandler.fs
namespace StockTicker.Commands

open System
open System.Collections.Generic
open StockTicker.Events
open EventStorage
open StockTicker.Core
open System.Threading.Tasks
open StockTicker
open StockMarket
open Events

// Listing 14.4 Command handler with async retry logic
module CommandHandler =
    // Listing 9.4 Retry Async Builder
    let private retryPublish = RetryAsyncBuilder(10, 250) // #A

    let tradingCoordinator = TradingCoordinator.Instance()   // #B
    let Storage = new EventStorage()     // #C

    let AsyncHandle (commandWrapper:CommandWrapper) =   // #D
        let connectionId = commandWrapper.ConnectionId

        retryPublish {      // #A
            tradingCoordinator.PublishCommand(PublishCommand(connectionId, commandWrapper))   // #E

            let event =
                let cmd = commandWrapper.Command
                match cmd with     // #F
                | BuyStockCommand(connId,trading) -> StocksBuyedEvent(commandWrapper.Id, trading)
                | SellStockCommand(connId, trading) -> StocksSoldEvent(commandWrapper.Id, trading)   // #F

            let eventDescriptor = Event.Create (commandWrapper.Id, event)
            Storage.SaveEvent (Guid(connectionId)) eventDescriptor    // #G
        }